osquery/tables/system/windows/objects.cpp

```diff
@@ -0,0 +1,214 @@
+  KObjHandle kdo;
+  if (!kdo.openDirObj(directory)) {
+    return objects;
+  }
+
+  // iterator index is incremented by NtQueryDirectoryObject
+  for (DWORD index = 0;;) {
+    BYTE rgDirObjInfoBuffer[1024 * 8] = {0};
+    POBJDIR_INFORMATION pObjDirInfo = (POBJDIR_INFORMATION)rgDirObjInfoBuffer;
+
+    //  get the name and type of the index'th object in the directory
+    NTSTATUS ntStatus = NtQueryDirectoryObject(kdo.getAsHandle(),
+                                               pObjDirInfo,
+                                               sizeof(rgDirObjInfoBuffer),
+                                               TRUE,
+                                               FALSE,
+                                               &index,
+                                               NULL);
+    if (STATUS_SUCCESS != ntStatus) {
+      break;
+    }
+
+    obj_name_type_pair object;
+    object.first = (pObjDirInfo->ObjectName.Buffer);
+    object.second = (pObjDirInfo->ObjectTypeName.Buffer);
+
+    objects.push_back(object);
+  }
+
+  return objects;
+}
+
+// enumerate all objects in a given windows terminal services session
+//
+// objects are found in the windows object directory
+// "\Sessions\BNOLINKS\<sessionnum>"
+//
+std::vector<obj_name_type_pair>
+EnumerateBaseNamedObjectsLinks(std::wstring session_num,
+                               std::wstring object_type) {
+  std::vector<obj_name_type_pair> objects;
+
+  // look up NtQuerySymbolicLinkObject as exported from ntdll
+  NTQUERYSYMBOLICLINKOBJECT NtQuerySymbolicLinkObject =
+      (NTQUERYSYMBOLICLINKOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
+                                                "NtQuerySymbolicLinkObject");
+  if (NULL == NtQuerySymbolicLinkObject) {
+    return objects;
+  }
+
+  // by convention, we expect there to be <n> objects in \Sessions\BNOLINKS with
+  // the following three characteristics:
+  //
+  //   (1) the object name is the string representation of of an active terminal
+  //       services session id.  this means we expect the object name to be a
+  //       string representation of an integer
+  //
+  //   (2) the object type name be "SymbolicLink"
+  //
+  //   (3) the symbolic link point to a directory object
+  //
+  // validate in this order
+  //
+
+  // validate (1)
+  //
+  // validate that this appears to be a valid terminal services session id
+  // another approach is to enumerate windows terminal services sessions with
+  // WTSEnumerateSessions and validate against that list
+  //
+  if (!(L"0" == session_num || safe_wstr_to_int(session_num) > 0)) {
+    return objects;
+  }
+
+  // validate (2)
+  //
+  // validate that the object type is "SymbolicLink"
+  //
+  if (L"SymbolicLink" != object_type) {
+    return objects;
+  }
+
+  // at this point we have SymbolicLink with a name matching a terminal services
+  // session id.  now build the fully qualified object path
+  std::wstring qualifiedpath = L"\\Sessions\\BNOLINKS\\" + session_num;
```

Comment:
Seems like some of these strings should be `const` at the global scope, might allow for some variable re-use and make constructing things a bit easier?

Reply:
Let's also make this `const auto`.